### PR TITLE
Add optional `serde` feature for `serde` support.

### PR DESCRIPTION
**Pull request by jthulhu**
This commit add a feature "serde" (disabled by default) which pulls in the optional dependency serde and implements `Serialize` and `Deserialize` for `IntervalTree`.

### CODE REVIEW

**Comment by jonathanGB** [comment]
Thanks @TheBlackBeans! Could you please add tests that validate that serialization/deserialization work? I would expect the following cases to be tested:

* `IntervalTree`: for both an empty tree and a populated tree of height >= 2
  * [optional] Serialization produces the expected output (e.g. expected size and expected children nodes)
  * Deserializing the serialization of a tree should be equal to the original tree
* `Node`: for both a leaf (i.e. no children) and an internal node (i.e. at least 1 child node)
  * [optional] Serialization produces the expected output (e.g. expected key and value)
  * Deserializing the serialization of a node should be equal to the original node

**Comment by jthulhu** [comment]
@jonathanGB done. I also added deserialization-only tests, both for empty and non-empty trees, and for leaf and internal nodes.

**Comment by jthulhu** [comment]
Ops I didn't read the condition height ≥ 2, I'll patch that ASAP.

**Comment by jthulhu** [comment]
Alright now it should fit your requirements.

File: Cargo.toml [hunk 1]
```diff
@@ -12,4 +12,9 @@ keywords = ["interval", "tree", "bound", "exclusive", "difference"]
 categories = ["algorithms", "data-structures"]
 
 [dependencies]
-rand = "0.8"
\ No newline at end of file
+rand = "0.8"
+serde = { version = "1.0.145", optional = true , features = ["derive"] }
```
Comment (jonathanGB) [inline Cargo.toml:16]: Is there a reason to select version "1.0.145" specifically? I think I would prefer to just specify "1.0", that is we don't limit to a specific patch number.

Comment (jonathanGB) [inline Cargo.toml]: Is there a reason why we have "1.0.145" specifically? Otherwise, I would prefer to just have "1.0", so that we do not limit to one patch number -- just like we do for `rand` above.

File: Cargo.toml [hunk 2]
```diff
@@ -12,4 +12,9 @@ keywords = ["interval", "tree", "bound", "exclusive", "difference"]
...
+
+[dev-dependencies]
+serde_json = "1.0"
+serde = { version = "1.0.145", features = ["derive"] }
```
Comment (jonathanGB) [inline Cargo.toml]: Same here.

File: src/node.rs [hunk 3]
```diff
@@ -1,9 +1,112 @@
 use std::fmt;
 use std::ops::Bound;
 use std::ops::Bound::*;
+#[cfg(any(feature="serde", test))]
+use serde::{Serialize, Deserialize};
 
 pub(crate) type Range<K> = (Bound<K>, Bound<K>);
 
+#[cfg(test)]
```
Comment (jonathanGB) [inline src/node.rs]: Could you please move the tests to the bottom of this file. I prefer to have the implementation before tests.

**Review by jonathanGB — CHANGES_REQUESTED** [review]
Thanks for adding the tests! Just a few nits that should be easy to resolve, and then we should be able to merge :)

File: Cargo.toml (hunk 1 above)
Comment (jthulhu) [reply Cargo.toml]: No particular reason, it's just what cargo-edit specified.

File: Cargo.toml (hunk 2 above)
Comment (jthulhu) [reply Cargo.toml]: See the other.

File: src/node.rs (hunk 3 above)
Comment (jthulhu) [reply src/node.rs]: As you wish!

**Comment by jthulhu** [comment]
There we go.

**Comment by jonathanGB** [comment]
Thanks!

**Comment by jonathanGB** [comment]
FYI I noticed that we forgot to add the `serde` feature in the Cargo.toml, thus users of the interval tree can't serialize/deserialize. It worked in the tests because the dependency is not optional in the `dev-dependencies`.

This is fixed in https://github.com/jonathanGB/Unbounded-Interval-Tree/commit/3a8146dd710bae9fff6ddecac4d0dff93e1e5afb, and the new version is published on crates.io under version 1.1.2: https://crates.io/crates/unbounded-interval-tree/1.1.2.

**Comment by jthulhu** [comment]
@jonathanGB you don't need to specify a feature in this case. If a dependency is optional, it will automatically create a feature with the same name of the dependency, which will only enable that dependency. This is why in the code `#[cfg(feature="serde")]` works without having to declare that feature. See [the reference of this](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies) in the Cargo Book.

I added it as a plain dependency in dev mode simply because it's a pain to launch tests with `--features serde` or similar stuff, and because my editor reports to me that some pieces of code are disabled if it has a `#[cfg(feature="...")]` and that the feature is not enabled by default.

**Comment by jthulhu** [comment]
Also, I noticed that in the latest version of this package, you have moved `UnboundedIntervalTree` inside the `interval_tree` module, which is a breaking change, because now you have to use `unbounded_interval_tree::interval_tree::IntervalTree` instead of `unbounded_interval_tree::IntervalTree`, so maybe a major version bump would be necessary? Besides that, it's perfect, thanks for the merging the pull!
